File: 999/999.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/999.py

val MOD: Long = 1_234_567_891L
val INV_TWO: Long = (MOD + 1L) / 2L

val SMALL_W: Array<Long> = arrayOf(0L, 1L, 2L, -4L, -32L, -192L, 3584L, 77824L, 262144L)

fun smallW(index: Long): Long {
    if (index < 0L) {
        val pos = smallW(-index)
        return (MOD - pos % MOD) % MOD
    }
    val v = SMALL_W[index.toInt()]
    return if (v < 0L) ((v % MOD) + MOD) % MOD else v % MOD
}

fun powMod(base: Long, exp: Long, modulus: Long): Long {
    var result = 1L
    var b = base % modulus
    var e = exp
    while (e > 0L) {
        if (e % 2L == 1L) {
            result = result * b % modulus
        }
        e /= 2L
        b = b * b % modulus
    }
    return result
}

fun edsBlock(n: Long): LongArray {
    if (n <= 4L) {
        return LongArray(8) { i -> smallW(n - 3L + i.toLong()) }
    }

    val middle = n / 2L
    val source = edsBlock(middle)
    val sourceStart = middle - 3L

    fun get(index: Long): Long = source[(index - sourceStart).toInt()]

    fun odd(index: Long): Long {
        val a = get(index + 1L)
        val b = powMod(get(index - 1L), 3L, MOD)
        val c = get(index - 2L)
        val d = powMod(get(index), 3L, MOD)
        return (a * b % MOD + MOD - c * d % MOD) % MOD
    }

    fun even(index: Long): Long {
        val a = get(index)
        val b = get(index + 2L)
        val c = powMod(get(index - 1L), 2L, MOD)
        val d = get(index - 2L)
        val e = powMod(get(index + 1L), 2L, MOD)
        val inner = (b * c % MOD + MOD - d * e % MOD) % MOD
        return a * INV_TWO % MOD * inner % MOD
    }

    return if (n % 2L == 0L) {
        longArrayOf(
            odd(middle - 1L),
            even(middle - 1L),
            odd(middle),
            even(middle),
            odd(middle + 1L),
            even(middle + 1L),
            odd(middle + 2L),
            even(middle + 2L)
        )
    } else {
        longArrayOf(
            even(middle - 1L),
            odd(middle),
            even(middle),
            odd(middle + 1L),
            even(middle + 1L),
            odd(middle + 2L),
            even(middle + 2L),
            odd(middle + 3L)
        )
    }
}

fun wMod(n: Long): Long {
    if (n < 0L) {
        val pos = wMod(-n)
        return (MOD - pos % MOD) % MOD
    }
    return edsBlock(n)[3]
}

fun aMod(n: Long): Long {
    require(n >= 1L)
    val rem = n % 4L
    val signPositive = rem == 1L || rem == 2L
    val exp = n * n / 4L
    val inverseScale = powMod(INV_TWO, exp, MOD)
    val wval = wMod(n)
    return if (signPositive) {
        wval * inverseScale % MOD
    } else {
        (MOD - wval * inverseScale % MOD) % MOD
    }
}

fun main() {
    check(aMod(1L) == 1L)
    check(aMod(2L) == 1L)
    check(aMod(3L) == 1L)
    check(aMod(4L) == 2L)
    check(aMod(13L) == 23321L)
    check(aMod(1003L) == 231906014L)

    println(aMod(1_000_000_000_000_000_003L))
}

File: 999/999.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/999.py

const MOD: u64 = 1_234_567_891;
const INV_TWO: u64 = (MOD + 1) / 2;

const SMALL_W: [i64; 9] = [0, 1, 2, -4, -32, -192, 3584, 77824, 262144];

fn small_w(index: i64) -> u64 {
    if index < 0 {
        let pos = small_w(-index);
        return (MOD - pos % MOD) % MOD;
    }
    let v = SMALL_W[index as usize];
    if v < 0 {
        ((v % MOD as i64 + MOD as i64) as u64) % MOD
    } else {
        (v as u64) % MOD
    }
}

fn pow_mod(mut base: u64, mut exp: u64, modulus: u64) -> u64 {
    let mut result = 1u64;
    base %= modulus;
    while exp > 0 {
        if exp % 2 == 1 {
            result = result * base % modulus;
        }
        exp /= 2;
        base = base * base % modulus;
    }
    result
}

fn eds_block(n: i64) -> [u64; 8] {
    if n <= 4 {
        let mut result = [0u64; 8];
        for i in 0..8 {
            result[i] = small_w(n - 3 + i as i64);
        }
        return result;
    }

    let middle = n / 2;
    let source = eds_block(middle);
    let source_start = middle - 3;

    let get = |index: i64| -> u64 {
        source[(index - source_start) as usize]
    };

    let odd = |index: i64| -> u64 {
        let a = get(index + 1);
        let b = pow_mod(get(index - 1), 3, MOD);
        let c = get(index - 2);
        let d = pow_mod(get(index), 3, MOD);
        (a * b % MOD + MOD - c * d % MOD) % MOD
    };

    let even = |index: i64| -> u64 {
        let a = get(index);
        let b = get(index + 2);
        let c = pow_mod(get(index - 1), 2, MOD);
        let d = get(index - 2);
        let e = pow_mod(get(index + 1), 2, MOD);
        let inner = (b * c % MOD + MOD - d * e % MOD) % MOD;
        a * INV_TWO % MOD * inner % MOD
    };

    if n % 2 == 0 {
        [
            odd(middle - 1),
            even(middle - 1),
            odd(middle),
            even(middle),
            odd(middle + 1),
            even(middle + 1),
            odd(middle + 2),
            even(middle + 2),
        ]
    } else {
        [
            even(middle - 1),
            odd(middle),
            even(middle),
            odd(middle + 1),
            even(middle + 1),
            odd(middle + 2),
            even(middle + 2),
            odd(middle + 3),
        ]
    }
}

fn w_mod(n: i64) -> u64 {
    if n < 0 {
        let pos = w_mod(-n);
        return (MOD - pos % MOD) % MOD;
    }
    eds_block(n)[3]
}

fn a_mod(n: i64) -> u64 {
    assert!(n >= 1);
    let sign_positive = n % 4 == 1 || n % 4 == 2;
    let exp = (n as u64) * (n as u64) / 4;
    let inverse_scale = pow_mod(INV_TWO, exp, MOD);
    let wval = w_mod(n);
    if sign_positive {
        wval * inverse_scale % MOD
    } else {
        (MOD - wval * inverse_scale % MOD) % MOD
    }
}

fn main() {
    assert_eq!(a_mod(1), 1);
    assert_eq!(a_mod(2), 1);
    assert_eq!(a_mod(3), 1);
    assert_eq!(a_mod(4), 2);
    assert_eq!(a_mod(13), 23321);
    assert_eq!(a_mod(1003), 231906014);

    println!("{}", a_mod(1_000_000_000_000_000_003));
}